module/signature/signer_indices.go

```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		return nil, fmt.Errorf("could not extract signer indices from prefixed data: %w", err)
+	}
+
+	numberCanonicalNodes := len(canonicalIdentities)
+	if e := validPadding(signerIndices, numberCanonicalNodes); e != nil {
+		return nil, fmt.Errorf("signerIndices padding are invalid: %w", e)
+	}
+
+	// decode bits to Identities
+	signerIdentities := make(flow.IdentityList, 0, numberCanonicalNodes)
+	for i := 0; i < numberCanonicalNodes; i++ {
+		if bitutils.ReadBit(signerIndices, i) == 1 {
+			signerIdentities = append(signerIdentities, canonicalIdentities[i])
+		}
+	}
+	return signerIdentities, nil
+}
+
+// validPadding verifies that `bitVector` satisfies the following criteria
+//  1. The `bitVector`'s length [in bytes], must be the _minimal_ possible length such that it can hold
+//    `numUsedBits` number of bits. Otherwise, we return an `ErrIncompatibleBitVectorLength`.
+//  2. If `numUsedBits` is _not_ an integer-multiple of 8, `bitVector` is padded with tailing bits. Per
+//     convention, these bits must be zero. Otherwise, we return an `ErrIllegallyPaddedBitVector`.
+// All errors represent expected failure cases for byzantine inputs. There are _no unexpected_ error returns.
+func validPadding(bitVector []byte, numUsedBits int) error {
+	// Verify condition 1:
+	l := len(bitVector)
+	if l != bitutils.MinimalByteSliceLength(numUsedBits) {
+		return fmt.Errorf("the bit vector contains a payload of %d used bits, so it should have %d bytes but has %d bytes: %w",
+			numUsedBits, bitutils.MinimalByteSliceLength(numUsedBits), l, ErrIncompatibleBitVectorLength)
+	}
+	// Condition 1 implies that the number of padded bits must be strictly smaller than 8. Otherwise, the vector
+	// could have fewer bytes and still have enough room to store `numUsedBits`.
+
+	// Verify condition 2, i.e. that all padded bits are all 0:
+	// * The number of padded bits are: `numPaddedBits := 8*len(bitVector) - len(canonicalIdentifiers)`
+	// * We know that numPaddedBits < 8; as `bitVector` passed check 1. Therefore, all padded bits are
+	//   located in `bitVector`s _last byte_.
+	// * Let `lastByte` be the last byte of `bitVector`. The leading bits, specifically `(8 - numPaddedBits)`
+	//   belong to the used payload, which could have non-zero values. We remove these using left-bit-shifts.
+	//   The result contains exactly all padded bits (plus some auxiliary 0-bits included by the bit-shift
+	//   operator). Hence, condition 2 is satisfied if and only if the result is identical to zero.
+	// Note that this implementation is much more efficient than individually checking the padded bits, as we check all
+	// padded bits at once; furthermore, we only use multiplication, subtraction, shift, which are fast.
+	if l == 0 {
+		return nil
+	}
+	// the above check has ensured that lastByte does exist
+	lastByte := bitVector[l-1]
+	numPaddedBits := 8*l - numUsedBits
+	if (lastByte << (8 - numPaddedBits)) != 0 {
```

Comment:
This seems to be equivalent and faster ( I haven't run the tests):
```suggestion
	if numUsedBits & 7 == 0 { // if numUsedBits is multiple of 8, then there are no padding bits to check
		return nil
	}
	// the above check has ensured that lastByte does exist (l==0 is excluded)
	lastByte := bitVector[l-1]
	if (lastByte << (numUsedBits & 7)) != 0 {   // shift by numUsedBits % 8
```

Reply:
Good idea. I verified with some tests, it is correct